File: Windows/Domain/StrategyModels.cs
using System.Text.Json.Serialization;

namespace CytisusTrading.Windows;

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum StrategySource
{
    Official,
    ThirdParty
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum StrategyRuntimeState
{
    Stopped,
    Starting,
    Ready,
    Running,
    Paused,
    Unhealthy,
    Exited,
    Rejected
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum ParameterType
{
    Number,
    Integer,
    Boolean,
    String
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum RiskTier
{
    Low,
    Medium,
    High
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum ParameterActivationMode
{
    Immediate,
    NextCycle,
    SafeBoundary
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum ParameterChangeResult
{
    Applied,
    PendingCycle,
    PendingConfirmation,
    PendingSafeBoundary,
    Rejected
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum LiveToPaperTransition
{
    StopOpeningRisk,
    Freeze,
    ControlledExit
}

public sealed record StrategyManifest(
    int SchemaVersion,
    int ProtocolVersion,
    string StrategyId,
    string Name,
    string Version,
    StrategySource Source,
    string Entrypoint,
    IReadOnlyList<StrategyMode> SupportedModes,
    IReadOnlyList<string> RequiredData,
    int ParameterSchemaVersion,
    string ParameterSchema,
    IReadOnlyList<string> RequestedCapabilities);

public sealed record ParameterSafetyOverride(
    string? HardMin,
    string? HardMax,
    RiskTier RiskTier,
    bool LiveMutable,
    bool RequiresPreview,
    bool RequiresPause,
    bool RequiresConfirmation,
    ParameterActivationMode ActivationMode,
    bool RollbackRequired);

public sealed record StrategyParameterDefinition(
    string Key,
    string Label,
    string Description,
    ParameterType Type,
    string DefaultValue,
    string? RecommendedMin,
    string? RecommendedMax,
    ParameterActivationMode RecommendedActivationMode,
    ParameterSafetyOverride SafetyOverride);

public sealed record StrategyParameterSchema(
    int SchemaVersion,
    string StrategyId,
    IReadOnlyList<StrategyParameterDefinition> Parameters);

public sealed record StrategyParameterChange(
    string ChangeId,
    string StrategyId,
    string ParameterKey,
    string OldValue,
    string NewValue,
    DateTimeOffset RequestedAt,
    DateTimeOffset? EffectiveAt,
    string RequestedBy,
    RiskTier RiskTier,
    int ParameterVersion,
    ParameterActivationMode ActivationMode,
    ParameterChangeResult Result,
    int RollbackVersion);

public sealed record StrategyMessageEnvelope(
    int SchemaVersion,
    string EventId,
    string CorrelationId,
    string StrategyId,
    string StrategyVersion,
    string CycleId,
    DateTimeOffset Timestamp,
    string MessageType,
    IReadOnlyDictionary<string, string> Payload);

public sealed record StrategySignal(
    string Symbol,
    double Score,
    string Reason,
    DateTimeOffset Timestamp);

public sealed record StrategyTarget(
    string Symbol,
    double TargetWeight,
    string Reason,
    DateTimeOffset Timestamp);

public sealed record StrategyIntentRecord(
    string IntentId,
    string Symbol,
    double TargetWeight,
    int Priority,
    bool AllowPartial,
    string ReasonCode,
    string CycleId,
    DateTimeOffset Timestamp,
    StrategyMode Mode);

public sealed record StrategyCycleSummary(
    string CycleId,
    DateTimeOffset CompletedAt,
    int SignalCount,
    int TargetCount,
    int IntentCount,
    string Result);

public sealed record StrategyPersistentState(
    string StrategyId,
    StrategyMode Mode,
    StrategyRuntimeState RuntimeState,
    HealthState Health,
    DateTimeOffset? LastHeartbeat,
    int ParameterVersion,
    IReadOnlyDictionary<string, string> ParameterValues,
    bool BlocksNewRisk,
    LiveToPaperTransition LiveToPaperTransition);

public sealed record LiveAuthorization(
    int SchemaVersion,
    string AuthorizationId,
    string StrategyId,
    IReadOnlyList<string> AllowedMarkets,
    IReadOnlyList<string> AllowedSymbolsOrUniverse,
    decimal MaximumCapital,
    decimal MaximumStrategyAllocation,
    decimal MaximumSinglePositionExposure,
    decimal MaximumDailyLoss,
    decimal MaximumDrawdown,
    int MaximumOrderFrequency,
    bool OutsideRegularHoursPermission,
    int ParameterVersion,
    DateTimeOffset ValidFrom,
    DateTimeOffset ExpiresAt,
    bool Enabled);

public sealed record ManifestValidationResult(
    bool IsValid,
    string Message)
{
    public static ManifestValidationResult Valid { get; } =
        new(true, "Manifest is valid.");

    public static ManifestValidationResult Invalid(string message)
    {
        return new ManifestValidationResult(false, message);
    }
}

public sealed record ModeSelectionResult(
    bool Accepted,
    StrategyMode Mode,
    string Message);

public sealed record ParameterChangeDecision(
    StrategyParameterChange Change,
    bool BlocksNewRisk,
    string ImpactPreview);

public sealed record PaperCycleResult(
    StrategyCycleSummary Cycle,
    DateTimeOffset LastHeartbeat,
    HealthState Health,
    IReadOnlyList<StrategySignal> Signals,
    IReadOnlyList<StrategyTarget> Targets,
    IReadOnlyList<StrategyIntentRecord> Intents,
    IReadOnlyList<StrategyMessageEnvelope> Messages);

public static class StrategyProtocolMessageTypes
{
    public static IReadOnlySet<string> CoreToStrategy { get; } =
        new HashSet<string>(StringComparer.Ordinal)
        {
            "initialize",
            "load_parameters",
            "apply_parameters",
            "market_snapshot",
            "allocation_budget",
            "start_cycle",
            "pause",
            "resume",
            "shutdown"
        };

    public static IReadOnlySet<string> StrategyToCore { get; } =
        new HashSet<string>(StringComparer.Ordinal)
        {
            "ready",
            "heartbeat",
            "log",
            "health",
            "signal",
            "factor_observation",
            "target_position",
            "trade_intent",
            "cycle_complete",
            "error"
        };
}

File: Sources/Domain/StrategyModels.swift
import Foundation

enum StrategySource: String, Codable, CaseIterable {
    case official = "Official"
    case thirdParty = "ThirdParty"
}

enum StrategyRuntimeState: String, Codable, CaseIterable {
    case stopped = "Stopped"
    case starting = "Starting"
    case ready = "Ready"
    case running = "Running"
    case paused = "Paused"
    case unhealthy = "Unhealthy"
    case exited = "Exited"
    case rejected = "Rejected"
}

enum ParameterType: String, Codable {
    case number = "Number"
    case integer = "Integer"
    case boolean = "Boolean"
    case string = "String"
}

enum RiskTier: String, Codable, CaseIterable {
    case low = "Low"
    case medium = "Medium"
    case high = "High"
}

enum ParameterActivationMode: String, Codable {
    case immediate = "Immediate"
    case nextCycle = "NextCycle"
    case safeBoundary = "SafeBoundary"
}

enum ParameterChangeResult: String, Codable {
    case applied = "Applied"
    case pendingCycle = "PendingCycle"
    case pendingConfirmation = "PendingConfirmation"
    case pendingSafeBoundary = "PendingSafeBoundary"
    case rejected = "Rejected"
}

enum LiveToPaperTransition: String, Codable, CaseIterable, Identifiable {
    case stopOpeningRisk = "StopOpeningRisk"
    case freeze = "Freeze"
    case controlledExit = "ControlledExit"

    var id: String { rawValue }
}

struct StrategyManifest: Codable, Identifiable, Equatable {
    let schemaVersion: Int
    let protocolVersion: Int
    let strategyId: String
    let name: String
    let version: String
    let source: StrategySource
    let entrypoint: String
    let supportedModes: [StrategyMode]
    let requiredData: [String]
    let parameterSchemaVersion: Int
    let parameterSchema: String
    let requestedCapabilities: [String]

    var id: String { strategyId }
}

struct ParameterSafetyOverride: Codable, Equatable {
    let hardMin: String?
    let hardMax: String?
    let riskTier: RiskTier
    let liveMutable: Bool
    let requiresPreview: Bool
    let requiresPause: Bool
    let requiresConfirmation: Bool
    let activationMode: ParameterActivationMode
    let rollbackRequired: Bool
}

struct StrategyParameterDefinition: Codable, Identifiable, Equatable {
    let key: String
    let label: String
    let description: String
    let type: ParameterType
    let defaultValue: String
    let recommendedMin: String?
    let recommendedMax: String?
    let recommendedActivationMode: ParameterActivationMode
    let safetyOverride: ParameterSafetyOverride

    var id: String { key }
}

struct StrategyParameterSchema: Codable, Equatable {
    let schemaVersion: Int
    let strategyId: String
    let parameters: [StrategyParameterDefinition]
}

struct StrategyParameterChange: Codable, Identifiable, Equatable {
    let changeId: String
    let strategyId: String
    let parameterKey: String
    let oldValue: String
    let newValue: String
    let requestedAt: Date
    let effectiveAt: Date?
    let requestedBy: String
    let riskTier: RiskTier
    let parameterVersion: Int
    let activationMode: ParameterActivationMode
    let result: ParameterChangeResult
    let rollbackVersion: Int

    var id: String { changeId }
}

struct StrategyMessageEnvelope: Codable, Identifiable, Equatable {
    let schemaVersion: Int
    let eventId: String
    let correlationId: String
    let strategyId: String
    let strategyVersion: String
    let cycleId: String
    let timestamp: Date
    let messageType: String
    let payload: [String: String]

    var id: String { eventId }
}

struct StrategySignal: Identifiable, Equatable {
    let symbol: String
    let score: Double
    let reason: String
    let timestamp: Date

    var id: String { "\(symbol)-\(timestamp.timeIntervalSince1970)" }
}

struct StrategyTarget: Identifiable, Equatable {
    let symbol: String
    let targetWeight: Double
    let reason: String
    let timestamp: Date

    var id: String { "\(symbol)-\(timestamp.timeIntervalSince1970)" }
}

struct StrategyIntentRecord: Codable, Identifiable, Equatable {
    let intentId: String
    let symbol: String
    let targetWeight: Double
    let priority: Int
    let allowPartial: Bool
    let reasonCode: String
    let cycleId: String
    let timestamp: Date
    let mode: StrategyMode

    var id: String { intentId }
}

struct StrategyCycleSummary: Identifiable, Equatable {
    let cycleId: String
    let completedAt: Date
    let signalCount: Int
    let targetCount: Int
    let intentCount: Int
    let result: String

    var id: String { cycleId }
}

struct StrategyPersistentState: Codable, Equatable {
    let strategyId: String
    var mode: StrategyMode
    var runtimeState: StrategyRuntimeState
    var health: HealthState
    var lastHeartbeat: Date?
    var parameterVersion: Int
    var parameterValues: [String: String]
    var blocksNewRisk: Bool
    var liveToPaperTransition: LiveToPaperTransition
}

struct LiveAuthorization: Codable, Identifiable, Equatable {
    let schemaVersion: Int
    let authorizationId: String
    let strategyId: String
    let allowedMarkets: [String]
    let allowedSymbolsOrUniverse: [String]
    let maximumCapital: Decimal
    let maximumStrategyAllocation: Decimal
    let maximumSinglePositionExposure: Decimal
    let maximumDailyLoss: Decimal
    let maximumDrawdown: Decimal
    let maximumOrderFrequency: Int
    let outsideRegularHoursPermission: Bool
    let parameterVersion: Int
    let validFrom: Date
    let expiresAt: Date
    let enabled: Bool

    var id: String { authorizationId }
}

struct ParameterChangeDecision {
    let change: StrategyParameterChange
    let blocksNewRisk: Bool
    let impactPreview: String
}

struct ModeSelectionResult {
    let accepted: Bool
    let mode: StrategyMode
    let message: String
}

struct PaperCycleResult {
    let cycle: StrategyCycleSummary
    let lastHeartbeat: Date
    let health: HealthState
    let signals: [StrategySignal]
    let targets: [StrategyTarget]
    let intents: [StrategyIntentRecord]
    let messages: [StrategyMessageEnvelope]
}

enum StrategyProtocolMessageTypes {
    static let coreToStrategy: Set<String> = [
        "initialize", "load_parameters", "apply_parameters",
        "market_snapshot", "allocation_budget", "start_cycle",
        "pause", "resume", "shutdown"
    ]
    static let strategyToCore: Set<String> = [
        "ready", "heartbeat", "log", "health", "signal",
        "factor_observation", "target_position", "trade_intent",
        "cycle_complete", "error"
    ]
}
